src/Network/Affjax/Request.purs

```diff
@@ -0,0 +1,115 @@
+  , ajax
+  ) where
+
+import Control.Monad.Aff (Aff(), EffA(), makeAff)
+import Control.Monad.Eff (Eff())
+import Control.Monad.Eff.Exception(Error())
+import Data.Array ()
+import Data.ArrayBuffer.Types (ArrayView())
+import Data.Function (Fn8(), runFn8)
+import Data.Maybe (Maybe(..), maybe)
+import Data.Nullable (Nullable(), toNullable)
+import DOM (Document())
+import DOM.File (Blob())
+import DOM.XHR (FormData())
+import Network.HTTP (Header(..))
+
+-- | The event type for AJAX requests.
+foreign import data Ajax :: !
+
+-- | The parameters for an AJAX request.
+type AjaxRequest a =
+  { url :: String
+  , method :: MethodName
+  , headers :: [Header]
+  , content :: Maybe (Content a)
+  , username :: Maybe String
+  , password :: Maybe String
+  }
+
+-- | A HTTP method name: `GET`, `POST`, etc.
+newtype MethodName = MethodName String
+
+-- | The types of data that can be set in an AJAX request.
+data Content a
+  = ArrayViewContent (ArrayView a)
+  | BlobContent Blob
+  | DocumentContent Document
+  | TextContent String
+  | FormDataContent FormData
+
+-- TODO: probably not this? Do we want to deal with other responses, include headers, etc?
+newtype AjaxResponse = AjaxResponse String
+
+-- | A basic request, `GET /` with no particular headers or credentials.
+defaultRequest :: forall c. AjaxRequest c
+defaultRequest =
+  { url: "/"
+  , method: MethodName "GET"
+  , headers: []
+  , content: Nothing
+  , username: Nothing
+  , password: Nothing
+  }
+
+-- | Make an AJAX request.
+ajax :: forall e a. AjaxRequest a -> Aff (ajax :: Ajax | e) AjaxResponse
+ajax req = makeAff $ runFn8
+  unsafeAjax req.url
+             (runMethodName req.method)
+             (runHeader <$> req.headers)
+             (toNullable $ runContent <$> req.content)
+             (toNullable req.username)
+             (toNullable req.password)
+  where
+  runMethodName (MethodName name) = name
+  runHeader (Header header value) = { header: show header, value: value }
+  runContent :: forall c. Content c -> XHRContent
+  runContent (ArrayViewContent av) = unsafeToXHRContent av
+  runContent (BlobContent b) = unsafeToXHRContent b
+  runContent (DocumentContent d) = unsafeToXHRContent d
+  runContent (TextContent s) = unsafeToXHRContent s
+  runContent (FormDataContent fd) = unsafeToXHRContent fd
+
+foreign import data XHRContent :: *
+
+foreign import unsafeToXHRContent
+  """
+  function unsafeToXHRContent (value) {
+    return value;
+  }
+  """ :: forall a. a -> XHRContent
+
+foreign import unsafeAjax
+  """
+  function unsafeAjax (url, method, headers, content, username, password, errback, callback) {
+    var xhr = new XMLHttpRequest();
+    xhr.open(method, url, true, username, password);
+    for (var i = 0, header; header = headers[i]; i++) {
+      xhr.setRequestHeader(header.header, header.value);
+    }
+    xhr.onerror = function (err) {
+      errback(err)();
+    };
+    xhr.onload = function () {
+      if (xhr.status === 200) callback(xhr.response)();
```

Comment:
Ah yep, good point. I can include that in the response record mentioned elsewhere in this thread.